libbeat/service/service_windows.go

```diff
@@ -71,10 +78,32 @@ func ProcessWindowsControlEvents(stopCallback func()) {
 	if isInteractive {
 		run = debug.Run
 	}
-	err = run(os.Args[0], &beatService{})
-	if err != nil {
-		logp.Err("Error: %v", err)
-	} else {
+
+	totalSleep := time.Duration(0)
+	for {
+		err = run(os.Args[0], &beatService{})
+		if err != nil {
+			// Sometimes it is the case that this is invoked against a service before that service has finished starting
+			if strings.Contains(err.Error, "The service process could not connect to the service controller") {
```

Comment:
Checking error strings should be a last resort when there is no other option because it's very brittle. But luckily in this case we have another option. It looks like you are testing for the occurrence of `ERROR_FAILED_SERVICE_CONTROLLER_CONNECT` which is errno 1063.

You'll have to trace through the code or do a test, but I expect that you can check to see if `err` is `syscall.Errno(1063)` (probably define your own constant if there isn't one already in the library for the error).

- https://docs.microsoft.com/en-us/windows/desktop/api/winsvc/nf-winsvc-startservicectrldispatchera
- https://msdn.microsoft.com/en-us/library/ms932980.aspx

Reply:
Thanks for the feedback, I agree, I hate using strings for the same reason. I didn't realize we could get that info from a syscall, I'll look into it.